src/app/core/ui/menu-item-list-editor/menu-item-list-editor.component.ts

```diff
@@ -0,0 +1,230 @@
+      }
+    });
+    return ids;
+  }
+
+  onDragDrop(event: CdkDragDrop<MenuItem[]>) {
+    try {
+      if (event.previousContainer === event.container) {
+        // Same container - always allow reordering
+        moveItemInArray(
+          event.container.data,
+          event.previousIndex,
+          event.currentIndex,
+        );
+      } else {
+        // Cross-container transfer - only allow if submenus are enabled
+        if (!this.allowSubMenu) {
+          // For shortcuts, prevent cross-container drops (no submenus allowed)
+          return;
+        }
+
+        transferArrayItem(
+          event.previousContainer.data,
+          event.container.data,
+          event.previousIndex,
+          event.currentIndex,
+        );
+
+        // Ensure the moved item has a subMenu array to allow nesting
+        const targetData = event.container.data as MenuItemForAdminUi[];
+        const movedItem = targetData[event.currentIndex];
+        if (movedItem && !movedItem.subMenu) {
+          movedItem.subMenu = [];
+        }
+
+        if (event.container.id === this.containerId) {
+          this.items = [...this.items];
+        }
+      }
+      this.emitItemsChange();
+    } catch (error) {
+      Logging.debug("Drag drop error:", error);
+    }
+  }
+
+  async addNewMenuItem() {
+    const newItem = new MenuItemForAdminUiNew(uuid());
+    const dialogRef = this.dialog.open(AdminMenuItemDetailsComponent, {
+      width: "600px",
+      data: {
+        item: { ...newItem },
+        isNew: true,
+        itemType: this.itemType,
+      },
+    });
+
+    dialogRef.afterClosed().subscribe((result) => {
+      if (result) {
+        const normalizedItem: MenuItemForAdminUi = {
+          ...result,
+          uniqueId: uuid(),
+          subMenu:
+            this.allowSubMenu && result.subMenu
+              ? MenuItemListEditorComponent.addUniqueIds(result.subMenu)
+              : [],
+        };
+        this.items = [normalizedItem, ...this.items];
+        this.emitItemsChange();
+      }
+    });
+  }
+
+  removeItem(item: MenuItemForAdminUi): void {
+    this.items = this.items.filter((i) => i !== item);
+    this.emitItemsChange();
+  }
+
+  onItemChange(newItem: MenuItemForAdminUi, index: number) {
+    this.items = [
+      ...this.items.slice(0, index),
+      newItem,
+      ...this.items.slice(index + 1),
+    ];
+    this.emitItemsChange();
+  }
+
+  /**
+   * Add unique IDs to all menu items for drag and drop functionality.
+   * This is necessary to ensure that each item can be uniquely identified
+   */
+  static addUniqueIds(items: MenuItem[]): MenuItemForAdminUi[] {
+    return items.map((item) => ({
+      ...item,
+      uniqueId: uuid(),
+      subMenu: item.subMenu
+        ? MenuItemListEditorComponent.addUniqueIds(item.subMenu)
+        : [],
+    }));
+  }
+
+  /**
+   * Convert MenuItemForAdminUi back to plain MenuItem by removing UI-specific properties
+   */
+  static toPlainMenuItem(
```

Comment:
Let's add some simple unit tests for `toPlainMenuItem` /items and `fromPlainMenuItem`/items (that'll probably boost the test coverage stats also - and will help identify any hidden problems).